### DISCUSSION TITLE
Adding a function to the python bindings

### DISCUSSION THREAD

**Discussion by carosto** [InteractiveComputerGraphics/SPlisHSPlasH#236] [Q&A] [answered]
Hello,

I am trying to create a Gym environment using the simulator in Python. For taking steps in the environment I need control over the execution of the single timesteps and cannot simply start/run the simulation using the runSimulation() function of the SimulatorBase.

To achieve this I want to use the timeStepNoGUI() function of SimulatorBase to take simulation steps manually. This function is included in the python bindings but for it to work, some final simulation initialization steps have to be taken. These steps are usually executed in the beginning of the runSimulation() function but since I am not using runSimulation(), I added a new function in SimulatorBase.cpp to replace this first part of runSimulation():

```
void SimulatorBase::finishInitialization()
{
	deferredInit();

	if (getStateFile() != "")
	{
		// avoid that command line parameter is overwritten
		const Real temp = m_stopAt;
		loadState(getStateFile());
		if (m_cmdLineStopAt)
			m_stopAt = temp;
		if (m_cmdLineNoInitialPause)
			m_doPause = false;
	}
	setCommandLineParameter();
}
```
I also added this function to the SimulatorBase.h file in the same way that the other functions are listed there (`void finishInitialization();`). 
Finally, to make the function callable in my python code I added it to the python bindings in SimulationModule.cpp for SimulatorBase. Again, I just copied the way that other functions were added there and switched the name for my function (`.def("finishInitialization", &SPH::SimulatorBase::finishInitialization)`).

I then built my modified version of SPlisHSPlasH using CMake and installed the python bindings as described in your installation instructions. This worked without issues.

However, when I include my newly written function finishInitialization() in my python programm, the programm keeps crashing with Windows fatal exception: access violation errors (under Windows) and Python Segmentation faults (under Linux). Interestingly, the error does not occur during the execution of the function, that part seems to work fine. It occurs at some (seemingly random) later point during the programms execution. Sometimes the programm runs for several timesteps without issue until it throws the error, other times it crashes before it even finishes the first timestep. 

Just running a simulation with runSimulation() works without issue, so the installation worked as expected. I assume that the issue is somewhere in the switching between Python and C++ but I do not know how to solve this. Have I missed any important steps for adding a function in the C++ code and transferring it to the python bindings?

Thanks in advance.

**Comment by janbender** [comment, marked as answer]
Hard to say where is the problem without testing it. 
Did you set: setUseGUI(false);
Otherwise the system will do some GUI callbacks without having a GUI. 

Can you post your Python script?

**Reply by carosto** [reply]
Yes, I've made sure not to use the GUI.

This is the relevant code for the simulation:
```
import numpy as np

from scipy.spatial.transform import Rotation as R

import os

import pysplishsplash as sph
import pysplishsplash.Utilities.SceneLoaderStructs as Scenes

import faulthandler; faulthandler.enable()

class Simulation:
    def __init__(self, use_gui=False,):

        self.cup_position = [0, 0, 0] 
        cup_rotation = R.from_euler('XYZ', [-90, 0, 0], degrees=True)
        self.cup_position.extend(cup_rotation.as_quat())

        self.jug_start_position = [0.5, 0, 0.5]
        jug_start_rotation = R.from_euler('XYZ', [90, 180, 0], degrees=True)
        self.jug_start_position.extend(jug_start_rotation.as_quat())

        self.output_directory = '../SimulationOutput'

        self.use_gui = use_gui

        self.cup_name = 'wine_glass_test'
        self.jug_name = 'beakernew'

        self.__init_simulation()

    def step(self):
        self.base.timeStepNoGUI()

    def __init_simulation(self):
        base = sph.Exec.SimulatorBase()
        base.init(
                useGui=self.use_gui, sceneFile=os.path.abspath('water_pouring/water_pouring/envs/scene.json'),
                stopAt=5,
                outputDir= self.output_directory
            )

        # Create an imgui simulator
        gui = sph.GUI.Simulator_GUI_imgui(base)
        base.setGui(gui)
        
        base.setTimeStepCB(self.__time_step_callback)

        # Get the scene and add objects
        scene = sph.Exec.SceneConfiguration.getCurrent().getScene()

        # Add the jug
        jug_file = f'{self.jug_name}.obj'
        rots = R.from_quat(self.jug_start_position[3:])
        r = np.linalg.norm(rots.as_rotvec())
        scene.boundaryModels.append(
                Scenes.BoundaryData(
                    meshFile=f'ObjectFiles/{jug_file}',
                    translation=self.jug_start_position[:3],
                    scale=[1, 1, 1],
                    color=[0.5, 0.5, 0.5, 1.0],
                    isAnimated=True,
                    axis=[1,0,0],
                    angle=r,
                    isWall=False,
                    mapInvert=False,
                    mapResolution=[25, 25, 25],
                )
            )

        # Add the cup
        cup_file = f'{self.cup_name}.obj'
        rots = R.from_quat(self.cup_position[3:])
        r = np.linalg.norm(rots.as_rotvec())
        scene.boundaryModels.append(
                Scenes.BoundaryData(
                    meshFile=f'ObjectFiles/{cup_file}',
                    translation=self.cup_position[:3],
                    scale=[1, 1, 1],
                    color=[0.5, 0.5, 0.5, 1.0],
                    isAnimated=True,
                    axis=[1,0,0],
                    angle=r,
                    isWall=False,
                    mapInvert=False,
                    mapResolution=[25, 25, 25],
                )
            )

        fluid_init_pos = np.array(self.jug_start_position[:3])
        fluid_init_pos_upper = fluid_init_pos + [0.07, 0.175, 0.07]

        scene.fluidBlocks.append(
                Scenes.FluidBlock(
                    id='Fluid',
                    boxMin=fluid_init_pos,
                    boxMax=fluid_init_pos_upper,
                    mode=0,
                    initialVelocity=[0.0, 0.0, 0.0],
                )
            )

        base.initSimulation()

        sim = sph.Simulation.getCurrent()
        
        sim.setValueInt(sim.BOUNDARY_HANDLING_METHOD, 0)

        base.readFluidParticlesState('water_pouring/water_pouring/envs/LiquidParticles/beakernew_fluid_block_p006_v5.bgeo', sim.getFluidModel(0))
            
        # move particles to the start position of the jug
        fluid_model = sim.getFluidModel(0)

        for i in range(fluid_model.numberOfParticles()):
            curr_pos = fluid_model.getPosition(i)
            fluid_model.setPosition(i, curr_pos + fluid_init_pos)
            fluid_model.setVelocity(i, [0.0, 0.0, 0.0])
            fluid_model.setAcceleration(i, [0.0, 0.0, 0.0])
        
        self.base = base
        self.base.finishInitialization()
        #self.base.runSimulation()

    def __time_step_callback(self):  
        sim = sph.Simulation.getCurrent()
        print('test')

        # keep the cup at a constant position
        boundary = sim.getBoundaryModel(1) 
        animated_body = boundary.getRigidBodyObject()  # second object is cup 
        animated_body.setPosition(self.cup_position[:3])
        animated_body.setRotation(self.cup_position[3:])
        animated_body.animate()    

if __name__ == "__main__":
  env = Simulation(use_gui=False)

  while True:
      print('start step')
      env.step()  
```

**Reply by janbender** [reply]
Why do you create a GUI in your code. I thought you wanted to have a simulation without GUI.

See:

```
        # Create an imgui simulator
        gui = sph.GUI.Simulator_GUI_imgui(base)
        base.setGui(gui)
```

**Reply by carosto** [reply]
Thank you, removing that part seems to have solved the issue. I ran previous simulations with the GUI and assumed that setting the useGui parameter in base.init() to FALSE would fully deactivate the GUI part here.

**Reply by janbender** [reply]
Great! 
I must take a closer look to find out what happens if you set a GUI and at the same time set the no-gui flag.
